README.md
# A Comparison of the CO2 Emissions From Automobiles Powered by Gasoline vs Electricity

## Abstract

Electric Vehicles are typically seen as an environmentally friendly alternative to traditional automobiles which emit Carbon Dioxide (CO2) in the process of burning gasoline.
Many Electric Vehicle manufacturers advertise their cars as "emission-free" because they emit no CO2 in the process of running, however, Electric Vehicles rely on their local power grid to charge their batteries, so an EV actually emits as much CO2 as their local energy generation sources. 

In order to verify the claim that Electric Vehicles are an environmental friendly alternative to traditional automobiles, the emissions from an Internal Combustion Engine (ICE) are compared with the emissions from the US power grid as a whole, as well as each individual state's power grid.

This investigation validates the claim that Electric Vehicles are an environmentally friendly alternative to traditional automobiles, discovering that **the average traditional autmoboile in the US emits 2.742 times more Carbon Dioxide than an Electric Vehicle**. 

As well, **the average Electric Vehicle emits less Carbon Dioxide than the average traditional automobile in all 50 states**, even states with heavy dependencies on fossil fuels.

Consequently, **Electric Vehicles in the United States offer substantial reductions to CO2 Emissions compared with traditional automobiles**, especially in states with a low dependence on fossil fuels.

## Source Code Investigation
Create and activate a virtual environment, then execute `requirements.sh`

* For the main comparisons of emissions between ICE vs EV, see Main.py
* To see the calculations for average CO2 emissions from an ICE, see CO2EmissionsFromICE.py
* To see the calculations for average CO2 emissions from the United States power grid, see TotalCO2EmissionsFromEV.py
* To see the calculations for average CO2 emissions from each state's power grid, see StateCO2EmissionsFromEV.py and 
* To see the results data and figures for state CO2 emissions, see StateResultsData/

# Methodology

In this investigation, publicly available data from the EIA is used to calculate and compare the amount of Carbon Dioxide (CO2) emitted from an Internal Combustion Engine (ICE) vs the power grid which supplies an Electric Vehicle (EV). 

Power sources will be compared using units of lbs CO2/mile driven, measuring the mass of Carbon Dioxide emitted while travelling one mile. For the purpose of this investigation, the emissions per mile are derived using averages and do not factor in speed or type of driving.

Although unrealistic, for the purposes of this investigation a lossless transfer of energy is assumed from the local power grid to Electric Vehicles, as well as a lossless transfer of energy for Internal Combustion Engines.

As well, it is assumed that 1 kWh of energy is utilized equally for Electric and Internal Combustion vehicles.

Gasoline used in this investigation is assumed to be a blend of approximately [10% ethanol content by volume](https://www.eia.gov/energyexplained/units-and-calculators/), and ethanol fuel is assumed to [generate equivalent emissions to finished motor gasoline](https://www.epa.gov/greenvehicles/greenhouse-gas-emissions-typical-passenger-vehicle#:~:text=typical%20passenger%20vehicle%3F-,A%20typical%20passenger%20vehicle%20emits%20about%204.6%20metric%20tons%20of,8%2C887%20grams%20of%20CO2). 

Battery mining/transportation process and gasoline treatment/transportation are out of the scope of this investigation, and therefore will be ignored.

This investigation does not account for EVs charged by personal solar panels, which would decrease CO2 emissions for a singular vehicle. 

Calculations for both types of vehicles can vary based on a multitude of factors, including make, model, year, size, and type of vehicle. For the purpose of this investigation both types of vehicles will be treated as singular entities using averages, however if the data were to be gathered then further work could be done to separate the calculations into different vehicle classes.

# Internal Combustion Engine (ICE)


The average Carbon Dioxide emissions from an ICE per mile driven can be calculated through a series of unit conversions. 

**Goal = ICE's CO2 emissions/Distance Travelled**

Starting with the three core pieces of input data, let
* A = CO2 emissions (kg)/gallon gasoline
* B = Energy generation (btu)/gallon gasoline
* C = Fuel Efficiency (mi)/gallon gasoline

and for unit conversions let
* x = lbs/kg
* y = btu/kWh

If
* Alpha = CO2 Emissions per Energy Generated
![](Data/ICEAlphaEquation.png)

and
* Beta = Miles Travelled per Energy Generated
![](Data/ICEBetaEquation.png)

Then
* **Goal = CO2 Emissions per Mile Travelled**
![](Data/ICEGoalEquation.png)


## Sources:

* [CO2 Emissions](https://www.epa.gov/greenvehicles/greenhouse-gas-emissions-typical-passenger-vehicle#:~:text=typical%20passenger%20vehicle%3F-,A%20typical%20passenger%20vehicle%20emits%20about%204.6%20metric%20tons%20of,8%2C887%20grams%20of%20CO2)
* [Energy Generation](https://www.eia.gov/energyexplained/units-and-calculators/)
* Conversions: [btu/kWh](https://www.eia.gov/energyexplained/units-and-calculators/energy-conversion-calculators.php) and [lbs/kg](https://www.rapidtables.com/convert/weight/kg-to-pound.html)

# Power Grid Supplying Electric Vehicles (EV)

Underlying this investigation is the assumption that Electric Vehicles are charged using electricity generated from the power grid, so the CO2 emitted from electric vehicles is equivalent to the average CO2 emitted from all power sources used to generate the electricity

The weighted average of CO2 emissions per kWh of energy generated is calculated from the product of each power source and the percent of total energy supplied.

In order to compare with ICE emissions, the average lbs CO2 emissions per kWh of energy generated is then converted to lbs CO2 emissions per Mile.

## Equations

**Goal = Power grid's CO2 emissions/Mile travelled**

Gamma = Power grid's CO2 emissions/kWh energy generated

Let 
* S = Energy Sources
* i = Each energy source
* C<sub>i</sub> = CO2 emissions (lbs/kWh) for energy source
* P<sub>i</sub> = Percent of total energy generated for energy source

Such that

![](Data/EVGammaEquation.png)

Then, let
* M = Average EV efficiency (Mi/kWh)

Such that

![](Data/EVGoalEquation.png)

The weighted average of lbs CO2/kWh and lbs CO2/kWh are calculated for the US as a whole, as well as each individual state.


## Emissions For Each Energy Source

The United States' power supply consists of some combination of Coal, Petroleum, Natural Gas, Other Gases, Other (various combustion based methods), Solar, Wind, Hydroelectric, Geothermal, and Biomass.

| Energy Source  | Emissions (lbs CO2/kWh) | Explanation                                                                                                                   |
|---------------|-------------------------|-------------------------------------------------------------------------------------------------------------------------------|
| Coal          | 2.230                   | Fossil Fuel                                                                                                                   |
| Petroleum     | 2.195                   | Fossil Fuel                                                                                                                   |
| Natural Gas   | 0.906                   | Fossil Fuel                                                                                                                   |
| Other Gases   | 0.906                   | Assumption: Emissions are equivalent to Natural Gas                                                                           |
| Other         | 1.777                   | Assumption: Emissions are equivalent to the fossil fuel average                                                               |
| Solar         | 0                       | Carbon Neutral                                                                                                                     |
| Wind          | 0                       | Carbon Neutral                                                                                                                     |
| Hydroelectric | 0                       | Carbon Neutral                                                                                                                     |
| Geothermal    | .017                    | Assumption: Emissions are approximately 1% of the fossil fuel average                                                                                                        |
| Nuclear       | 0                       | Assumption: Disregard possible emissions required to acquire nuclear fuel                                                     |
| Biomass       | 0                       | Although burning biological fuel emits CO2, the EIA considers biomass to be carbon neutral because of the carbon absorption from biological organisms|


## Sources:

* [Fossil Fuels](https://www.eia.gov/tools/faqs/faq.php?id=74&t=11)
* [Renewable](https://www.eia.gov/tools/faqs/faq.php?id=74&t=11)
* [Geothermal](https://www.eia.gov/energyexplained/geothermal/geothermal-energy-and-the-environment.php)
* [Nuclear](https://www.eia.gov/energyexplained/nuclear/nuclear-power-and-the-environment.php)
* [Biomass](https://www.eia.gov/energyexplained/biomass/biomass-and-the-environment.php)
* 

## Energy Generation Source Data

* [United States Power Grid](https://www.eia.gov/tools/faqs/faq.php?id=427&t=3) from 2021

* [State Power Grid](https://www.eia.gov/electricity/data/state/) from 2021


# Analysis:
An Internal Combustion Engine burning traditional finished motor gasoline emits an average of **0.809 lbs CO2/kWh**, while the United States power grid as a whole produces an average of **0.295 lbs CO2/kWh** from various power sources.
Thus, **the average Electric Vehicle in the US only emits about 36% as much Carbon Dioxide as a traditional automobile**.

However, no Electric Vehicles are powered from the entire US power grid, but rather each vehicle depends on its local power grid. 

Therefore, classifying electricity generation by state helps identify the areas of the country where an Electric Vehicle can charge with the lowest emissions.

![](StateResultsData/NormalizedStateEmissionsPerMile.png)

### Findings:
* Electric Vehicles emit less CO2 per mile driven than a traditional car in all 50 states.

* The states with the lowest emissions all utilize some combination of nuclear and renewable energy sources.

* While New York, New Jersey, and Connecticut all generate at least half of their energy from fossil fuels, they still rank 8th, 10th and 11th respectively in lowest CO2 emissions by using nuclear and other renewable energy sources.

* Hydroelectric power is closely associated with low CO2 emissions, with 7 of the 10 lowest emitting states generating at least 20% of their energy using Hydroelectric methods. The 3 lowest emitting states all generate at least 48% of their energy using Hydroelectric methods.

* On the contrary, coal use is closely associated with high CO2 emissions, with 7 of the 8 highest emitting states generating at least 50% of their energy from coal. The only exception is Hawaii, the 3rd highest emitting state, which generates 67% of its energy from petroleum, the only state generating at least 15% of its energy using petroleum.

* West Virginia, the state with the highest CO2 emissions, still emits 12% less CO2 than the average ICE automobile.

# Conclusions
Although Electric Vehicles in the United States are not emission free, they provide significant reductions in Carbon Dioxide emissions compared to traditional Internal Combustion Engine Vehicles nationwide. 

While CO2 emissions vary dramatically among all states, even the states with the heaviest reliance on fossil fuels for energy still find a moderate reduction in emissions. 

**Therefore, Electric Vehicles can be confidently deemed an environmentally friendly alternative to tradition gasoline powered automobiles.**



## Options For Continuing Research
* Integrating small scale photovoltaic systems (household/business solar panels)
* Applying advanced analytical techniques beyond my current level of training to the state source generation data
* Correlation between fossil fuel reliance and CO2 emissions (compare with ICE threshold and potentially a predictive CNN)
* Increase in energy production required to support increased usage of EVs
* Percent of EV charging by personal solar panels required to equal ICE emissions per state
* Incorporate EV battery mining and transportation into emissions model
* Incorporate gasoline treatment and transportation into emissions model
* Compare diesel powered trucking fleets to theoretical EV equivalent


changelog.md
Initial version: released November 2022

V2: released March 2023:
- Initial version was built on the faulty implicit assumption that ICE automobiles and EVs utilize energy at the same rate by comparing emissions per kWh.
- To account for the fact that the energy generated is harnessed in different ways, Electric Vehicles and Gasoline Vehicles are now compared with emissions per mile driven
- Calculations, results data, and documentation have all been updated to reflect this change.

CO2EmissionsFromICE.py
# CO2 Emissions from 1 gallon of gasoline can vary depending on the fuel efficiency and car model
# Assumption: We are comparing EV to traditional ICE using gasoline fuel
# Average kg CO2 emitted from burning one gallon of gasoline
avg_kg_CO2_per_gallon_gas = 8.887

lbs_per_kg = 2.205

avg_lbs_CO2_per_gallon_gas = round(avg_kg_CO2_per_gallon_gas * lbs_per_kg , 3)

# Energy produced from 1 gallon of finished motor gasoline (containing about 10% fuel ethanol by volume), measured in Btu (British Thermal Units)
avg_btu_per_gallon_gas = 120238

btu_per_kwh = 3412

avg_kwh_per_gallon_gas = round(avg_btu_per_gallon_gas / btu_per_kwh, 3)

# CO2/kWh = C02/gallon gas * gallon gas/kWh
avg_lbs_CO2_per_kwh = round(avg_lbs_CO2_per_gallon_gas / avg_kwh_per_gallon_gas, 3)

# Average miles per gallon for cars
# https://afdc.energy.gov/data/10310
avg_miles_per_gallon = 24.2
avg_miles_per_kwh = round(avg_miles_per_gallon / avg_kwh_per_gallon_gas, 3)
avg_lbs_CO2_per_mile = round(avg_lbs_CO2_per_kwh / avg_miles_per_kwh, 3)

if __name__ == "__main__":
    print(f"Average CO2 emitted from an ICE burning 1 gallon of gasoline: {avg_lbs_CO2_per_gallon_gas} lbs")
    print(f"Average amount of energy generated from 1 gallon of gasoline: {avg_kwh_per_gallon_gas} kWh")
    print(f"Average CO2 emitted per kWh generated from an ICE burning gasoline: {avg_lbs_CO2_per_kwh} lbs")
    print(f"Average CO2 emitted per mile traveled using an ICE burning gasoline: {avg_lbs_CO2_per_mile} lbs")

FossilFuelsCO2PerKWH.py
import pandas as pd

electricity_and_co2_by_source = pd.read_csv("Data/PowerAndCO2EmissionsBySource.csv")
lbs_per_kwh = {}


coal = electricity_and_co2_by_source.loc[electricity_and_co2_by_source["Sources"] == "Coal"]
# Unit: Million kWh
coal_electricity_generation = int(coal["Electricity Generation (million kWh)"])
# Unit: Million US tons
coal_CO2_produced = int(coal["CO2 Emissions (million US tons)"])
# Unit: lbs/kWh
lbs_per_kwh["Coal"] = round((coal_CO2_produced*2000)/coal_electricity_generation, 3)


nat_gas = electricity_and_co2_by_source.loc[electricity_and_co2_by_source["Sources"] == "Natural Gas"]
# Unit: Million kWh
nat_gas_electricity_generation = int(nat_gas["Electricity Generation (million kWh)"])
# Unit: Million US tons
nat_gas_CO2_produced = int(nat_gas["CO2 Emissions (million US tons)"])
# Unit: lbs/kWh
lbs_per_kwh["Natural Gas"] = round((nat_gas_CO2_produced*2000)/nat_gas_electricity_generation, 3)


petroleum = electricity_and_co2_by_source.loc[electricity_and_co2_by_source["Sources"] == "Petroleum"]
# Unit: Million kWh
petroleum_electricity_generation = int(petroleum["Electricity Generation (million kWh)"])
# Unit: Million US tons
petroleum_CO2_produced = int(petroleum["CO2 Emissions (million US tons)"])
# Unit: lbs/kWh
lbs_per_kwh["Petroleum"] = round((petroleum_CO2_produced*2000)/petroleum_electricity_generation, 3)

if __name__ == "__main__":
    print(lbs_per_kwh)

TotalCO2EmissionsFromEV.py
import pandas as pd
import FossilFuelsCO2PerKWH
from copy import deepcopy

# The average electric car kWh per 100 miles (kWh/100 mi) is 34.6.
# This is based on 231 electric cars, built between 2000 and 2022, and their kWh/100 mi as stated on fueleconomy.gov (the official US government source for fuel economy information).
# https://ecocostsavings.com/average-electric-car-kwh-per-mile/
avg_EV_miles_per_kwh = 1/.346

### Prepare dictionary mapping power source: lbs CO2 emitted per kwh energy generated
EV_lbs_per_kwh = deepcopy(FossilFuelsCO2PerKWH.lbs_per_kwh)

# Renewable energy sources generate no CO2 emissions
# Assumption: Nuclear generates no CO2, ignoring possible emissions during supplies extraction for simplicity
# Assumption: The plants that are the source of biomass for energy capture almost the same amount of CO2 through photosynthesis while growing as is released when biomass is burned, which can make biomass a carbon-neutral energy source
for renewable in ["Nuclear", "Wind", "Hydropower", "Solar", "Biomass", ]:
    EV_lbs_per_kwh[renewable] = 0

# Assumption: "Other gases" generate same emissions as natural gases
EV_lbs_per_kwh["Other gases"] = EV_lbs_per_kwh["Natural Gas"]
# Assumption: "Other sources" generates emissions equivalent to the average of all other fossil fuel sources
EV_lbs_per_kwh["Other sources"] = sum([EV_lbs_per_kwh[fossil_fuel] for fossil_fuel in ["Coal", "Natural Gas", "Petroleum"]])/3
# Assumption: "Geothermal" generates approximately 1% of fossil fuel emissions
EV_lbs_per_kwh["Geothermal"] = EV_lbs_per_kwh["Other sources"]/100


### Compute average CO2 emitted per kWh among all power sources
electricity_generation_by_source = pd.read_csv("Data/PowerGenerationBySource.csv")
fuel_sources = ["Natural Gas", "Coal", "Petroleum", "Other gases", "Nuclear", "Wind", "Hydropower", "Solar", "Biomass", "Geothermal", "Other sources"]
percent_electricity_generation = {}
for fuel_source in fuel_sources:
    percent_electricity_generation[fuel_source] = float(electricity_generation_by_source.loc[electricity_generation_by_source["Fuel Sources"] == fuel_source]["Percent Electricity Generation"])


us_avg_lbs_per_kwh = 0
for fuel_source in fuel_sources:
    us_avg_lbs_per_kwh += EV_lbs_per_kwh[fuel_source] * percent_electricity_generation[fuel_source]
us_avg_lbs_per_kwh = round(us_avg_lbs_per_kwh, 3)

### Calculate average miles per kwh to measure CO2 emissions per mile
us_avg_lbs_per_miles = round(us_avg_lbs_per_kwh / avg_EV_miles_per_kwh, 3)

if __name__ == "__main__":
    print("Fuel Source : Percent Total US Power Generation")
    for fuel_source, peg in percent_electricity_generation.items():
        print(f"{fuel_source} : {round(peg*100, 3)}%")
    print("\n")

    print("Fuel Source : CO2 (lbs) emitted in generating 1 kWh of power")
    for fuel_source, lbs_kwh in EV_lbs_per_kwh.items():
        if lbs_kwh == 0:
            print(f"{fuel_source} : Carbon Neutral")
        else:
            print(f"{fuel_source} : {round(lbs_kwh, 3)} lbs CO2/kWh")
    print("\n")

    print(f"CO2 emissions in the US to generate 1 kWh of power: {us_avg_lbs_per_kwh} lbs, averaged among all energy production sources")

    print(f"CO2 emissions in the US to travel 1 mile using the average Electric Vehicle: {us_avg_lbs_per_miles} lbs")

StateCO2EmissionsFromEV.py
import pandas as pd
import FossilFuelsCO2PerKWH
from copy import deepcopy
import plotly.express as px
from CO2EmissionsFromICE import avg_lbs_CO2_per_kwh as avg_ICE_CO2_per_kwh
from CO2EmissionsFromICE import avg_lbs_CO2_per_mile as avg_ICE_CO2_per_mile

state_EV_lbs_per_kwh = deepcopy(FossilFuelsCO2PerKWH.lbs_per_kwh)

# The average electric car kWh per 100 miles (kWh/100 mi) is 34.6.
# This is based on 231 electric cars, built between 2000 and 2022, and their kWh/100 mi as stated on fueleconomy.gov (the official US government source for fuel economy information).
# https://ecocostsavings.com/average-electric-car-kwh-per-mile/
avg_EV_miles_per_kwh = 1/.346

# Assumption: "Other gases" generate same emissions as natural gases
state_EV_lbs_per_kwh["Other Gases"] = state_EV_lbs_per_kwh["Natural Gas"]
# Assumption: "Other" generates emissions equivalent to the average of all other fossil fuel sources
state_EV_lbs_per_kwh["Other"] = sum([state_EV_lbs_per_kwh[fossil_fuel] for fossil_fuel in ["Coal", "Natural Gas", "Petroleum"]])/3
# Assumption: "Geothermal" generates approximately 1% of fossil fuel emissions
state_EV_lbs_per_kwh["Geothermal"] = state_EV_lbs_per_kwh["Other"]/100

# Assumption: "Other Biomass" and "Wood and Wood Derived Fuels" are both considered biomass, which is treated as Carbon Neutral by the EIA
# Renewable energy sources generate no CO2 emissions
# Assumption: Nuclear generates no CO2, ignoring possible emissions during supplies extraction for simplicity
for renewable in ["Hydroelectric Conventional", "Solar Thermal and Photovoltaic", "Other Biomass", "Wind", "Wood and Wood Derived Fuels", "Nuclear"]:
    state_EV_lbs_per_kwh[renewable] = 0


# Source: Net Generation by State by Type of Producer by Energy Source 
# https://www.eia.gov/electricity/data/state/
state_power = pd.read_csv("Data/AnnualStatePowerGeneration.csv")
state_power = state_power.dropna(axis=1, how="all")

state_total_power_in_2021 = state_power.loc[(state_power["YEAR"] == 2021) & (state_power["TYPE OF PRODUCER"] == "Total Electric Power Industry")]

states = ["AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DC", "DE", "FL", "GA", "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD", "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY"]

energy_sources = ['Solar Thermal and Photovoltaic', 'Wind', 'Other Biomass', 'Coal', 'Nuclear', 'Wood and Wood Derived Fuels', 'Hydroelectric Conventional', 'Other Gases', 'Other', 'Natural Gas', 'Petroleum', 'Geothermal', 'Total']

cols = ["LBS_CO2_PER_KWH", "LBS_CO2_PER_MILE"]
cols.extend(energy_sources)
output_df = pd.DataFrame(columns = cols, index=states)

### Compute average CO2 emitted per kWh among all power sources per state
for state in states:
    # Calculate percent power generation for each power type
    state_df = total_power_generation = state_total_power_in_2021.loc[(state_total_power_in_2021["STATE"] == state)]

    total_power_generation = int(state_df.loc[state_df["ENERGY SOURCE"] == "Total"]["GENERATION (Megawatthours)"])

    row = {"LBS_CO2_PER_KWH": 0, "LBS_CO2_PER_MILE": 0}

    state_lbs_per_kwh = 0
    total = 0
    for source in energy_sources:
        # check if the source exist in the state
        if source not in state_df["ENERGY SOURCE"].tolist():
            row[source] = 0
            continue
        elif source == "Total":
            continue
        source_power_generation = int(state_df.loc[state_df["ENERGY SOURCE"] == source]["GENERATION (Megawatthours)"])
        source_percent_generation = round(source_power_generation/total_power_generation, 3)
        
        row[source] = source_percent_generation
        total += source_percent_generation

        state_lbs_per_kwh += source_percent_generation * state_EV_lbs_per_kwh[source]
    
    row["Total"] = total
    row["LBS_CO2_PER_KWH"] = round(state_lbs_per_kwh, 3)
    row["LBS_CO2_PER_MILE"] = round(state_lbs_per_kwh / avg_EV_miles_per_kwh, 3)
    output_df.loc[state] = pd.Series(row)


if __name__ == "__main__":
    output_df.to_csv("StateResultsData/AllStates.csv")

    sorted_df = output_df.sort_values("LBS_CO2_PER_KWH")
    sorted_df.to_csv("StateResultsData/OrderedByEmissions.csv")

    states_with_less_emissions_per_mile = sorted_df[sorted_df["LBS_CO2_PER_MILE"] < avg_ICE_CO2_per_mile]
    states_with_less_emissions_per_mile.to_csv("StateResultsData/StatesWithLessEmissionsPerMileThanICE.csv")

    states_with_more_emissions_per_mile = sorted_df[sorted_df["LBS_CO2_PER_MILE"] >= avg_ICE_CO2_per_mile]
    states_with_more_emissions_per_mile.to_csv("StateResultsData/StatesWithMoreEmissionsPerMileThanICE.csv")

    # Emissions per kwh
    normalized_state_emissions = output_df.copy()
    normalized_state_emissions["Normalized CO2 Emissions"] = normalized_state_emissions["LBS_CO2_PER_KWH"].apply(lambda x: x - avg_ICE_CO2_per_kwh)
    normalized_state_emissions["States"] = normalized_state_emissions.index


    fig = px.choropleth(normalized_state_emissions,
                    locations='States', 
                    locationmode="USA-states", 
                    scope="usa",
                    color='Normalized CO2 Emissions',
                    color_continuous_scale=px.colors.diverging.PRGn_r,
                    color_continuous_midpoint=0,
                    title="Normalized CO2 Emissions (lbs CO2/kWh): Electric Vehicle (EV) Emissions Per State - Average Gasoline Vehicle (ICE) Emissions",
                    width=1200,
                    height=800,
                    )

    fig.update_layout( 
        # customize legend orientation & position
        legend=dict(
            title='Normalized CO2 Emissions Per kWh', orientation = 'h', y=1, yanchor="bottom", x=0.5, xanchor="center"
        )
    )
    fig.add_annotation(text="Darker means higher CO2 Emissions, Negative value means EVs emit less than ICE", y=0, showarrow=False)
    
    fig.write_image("StateResultsData/NormalizedStateEmissionsPerKWH.png")

    # Emissions per mile
    normalized_state_emissions = output_df.copy()
    normalized_state_emissions["Normalized CO2 Emissions"] = normalized_state_emissions["LBS_CO2_PER_MILE"].apply(lambda x: x - avg_ICE_CO2_per_mile)
    normalized_state_emissions["States"] = normalized_state_emissions.index


    fig = px.choropleth(normalized_state_emissions,
                    locations='States', 
                    locationmode="USA-states", 
                    scope="usa",
                    color='Normalized CO2 Emissions',
                    color_continuous_scale=px.colors.sequential.Blues,
                    title="Normalized CO2 Emissions (lbs CO2/mile): Electric Vehicle (EV) Emissions Per State - Average Gasoline Vehicle (ICE) Emissions",
                    width=1200,
                    height=800,
                    )

    fig.update_layout( 
        # customize legend orientation & position
        legend=dict(
            title='Normalized CO2 Emissions Per Mile', orientation = 'h', y=1, yanchor="bottom", x=0.5, xanchor="center"
        )
    )
    fig.add_annotation(text="Darker means higher CO2 Emissions, Negative value means EVs emit less than ICE", y=0, showarrow=False)
    
    fig.write_image("StateResultsData/NormalizedStateEmissionsPerMile.png")





